cmd/microcloud/preseed.go

```diff
@@ -903,31 +903,39 @@ func (p *Preseed) Parse(s *service.Handler, c *initConfig, installedServices map
 		c.systems[peer] = system
 	}
 
-	allResources := map[string]*lxdAPI.Resources{}
-	for peer, system := range c.systems {
-		// Skip any systems that had direct configuration.
-		if len(system.MicroCephDisks) > 0 || len(system.TargetStoragePools) > 0 || len(system.StoragePools) > 0 {
-			continue
+	var checkFilterZFS, checkFilterCeph bool
+	for _, system := range p.Systems {
+		if checkFilterZFS && checkFilterCeph {
+			break
 		}
 
-		setupStorage := false
-		for _, cfg := range system.JoinConfig {
-			if cfg.Entity == "storage-pool" {
-				setupStorage = true
-				break
-			}
+		if (DirectStorage{} == system.Storage.Local) {
+			checkFilterZFS = true
 		}
 
-		if setupStorage {
-			continue
+		if len(system.Storage.Ceph) == 0 {
+			checkFilterCeph = true
 		}
+	}
 
-		cert := system.ServerInfo.Certificate
+	if !checkFilterCeph && len(p.Storage.Ceph) > 0 {
+		return nil, fmt.Errorf("Ceph disk filter cannot be used. All systems have explicitly specified disks")
+	}
 
-		// Fetch system resources from LXD to find disks if we haven't directly set up disks.
-		allResources[peer], err = s.Services[types.LXD].(*service.LXDService).GetResources(context.Background(), peer, system.ServerInfo.Address, cert)
-		if err != nil {
-			return nil, fmt.Errorf("Failed to get system resources of peer %q: %w", peer, err)
+	if !checkFilterZFS && len(p.Storage.Local) > 0 {
+		return nil, fmt.Errorf("Local disk filter cannot be used. All systems have explicitly specified a disk")
+	}
+
+	allResources := map[string]*lxdAPI.Resources{}
+	if checkFilterZFS || checkFilterCeph {
```

Comment:
If you only check here if any of `checkFilterZFS` or `checkFilterCeph` is `true` then you will also fetch disks for systems which have explicitly defined disks.

I assume that is the error you are seeing in the preseed suite.

Reply:
Hmm, right, I've switched these to maps now, so we can selectively compile resources, and I've split the resource maps to be separate for ZFS and Ceph.